### DISCUSSION TITLE
How to handle bindgen ${name}-sys style crates?

### DISCUSSION THREAD

**Discussion by Avi-D-coder** [ipetkov/crane#200] [Q&A] [answered]
I'm trying to use `whisper-rs`/`whisper-rs-sys` as a dependency, it builds in my projects shell, it builds with vendored deps via `cargo vendor` in the shell, but nix build fails.

The error is reproduced directly on `whisper-rs` with the minimal flake in this fork  [whisper-rs/tree/flake](https://github.com/Avi-D-coder/whisper-rs/tree/flake).

`build.rs` is running `make` in the `whisper.cpp` directory.
```
last 10 log lines:
       >   cargo:rerun-if-changed=/nix/store/mmmjp0xcmr0dm307jv1l34hvi95z6d8s-glibc-2.35-224-dev/include/bits/types.h
       >   cargo:rerun-if-changed=/nix/store/78ywcf25bhbbfpv91wqirg4bd9dh3495-clang-wrapper-11.1.0/resource-root/include/stdbool.h
       >   gcc  -I.              -O3 -std=c11   -fPIC -pthread -mavx -mavx2 -mfma -mf16c   -c ggml.c -o ggml.o
       >
       >   --- stderr
       >   Assembler messages:
       >   Fatal error: can't create ggml.o: Permission denied
       >   make: *** [Makefile:145: ggml.o] Error 1
       >   thread 'main' panicked at 'Failed to build libwhisper.a', /nix/store/b21i5fqrmyrvmf7jnpbvnl4b4db3vdl3-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/whisper-rs-sys-0.2.1/build.rs:55:9
       >   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
 ```

**Comment by ipetkov** [comment, marked as answer]
Hi @Avi-D-coder! The problem is with symlinking whisperCpp.src because the build script tries to cd into that directory before invoking make. But with the symlink it ends up in the Nix store which is read only. 

You have a couple of options:
- Don't symlink the source but do a regular `cp -r`. You could probably get away with symlinking the regular files but the root directory will have to be writable (without changing the build.rs script that is)
- skip the source filtering completely, though any changes to the flake file (for example) will cause rebuilds
- write a source filter which includes the whisperCpp source. This could be a [composition filter](https://github.com/ipetkov/crane/blob/98894bb39b03bfb379c5e10523cd61160e1ac782/docs/API.md#libfiltercargosources) or you could go with something simpler like `lib.sourceByRegex`

Hope this helps!

**Comment by Avi-D-coder** [comment]
Thanks, the `whisper-rs` flake now builds, but I'm not clear, how this can be applied outside of the `whisper-rs` flake, where would I apply this when `whisper-rs` is a dependency.

@yusdacra  said:
> To fix this in dream2nix we basically set `cargoVendorDir = "$TMPDIR/nix-vendor"`; and then copy the vendor dir there in dependencies derivation and make it writable (`cp -rs --no-preserve=mode,ownership`) and in a `postInstall` we just move that vendor dir to `$out`, and in the main derivation we use that vendor directory from dependencies derivation out copying it in the same manner. This allows any dependencies `build.rs` to write stuff to their own source.

Here is the code: https://github.com/nix-community/dream2nix/blob/39c64b454e5f70e1a7b3c69dc595ee71f2d224e4/src/subsystems/rust/builders/crane/default.nix#L47

This feels like it should be the default given how common build.rs, and bindgen usage is.

**Reply by ipetkov** [reply]
> This feels like it should be the default given how common build.rs, and bindgen usage is

Ah but the error here has nothing to do with using bindgen, or using a build script (in fact, there are plenty of `*-sys` crates which build just fine and are part of the test suite). 

The problem is that _this particular build script_ [insists on building inside of the source directory](https://github.com/tazz4843/whisper-rs/blob/b81c6c94544ff9cf5157962ff84684fdb508f9eb/sys/build.rs#L49-L53) instead of having clean separation between source files and output artifacts (i.e. what cargo does and why artifacts go in the `target` directory and not right next to source files). The fact that the build script [moves the artifacts after the fact and then cleans up after itself](https://github.com/tazz4843/whisper-rs/blob/b81c6c94544ff9cf5157962ff84684fdb508f9eb/sys/build.rs#L57-L67) is another indicator of this code smell.

IMO this specific issue should be fixed upstream in the `whisper-sys` crate to be more friendly when the source directory is not writable. The [`cc` crate](https://crates.io/crates/cc) might be a good option to explore

---

As far as changing the default behavior goes, I feel that recursively copying all vendored sources into a writable directory is a needless pessimization for the majority of projects which build just fine with read-only sources. Keep in mind that this would have to be done on every single derivation which wants to run a cargo based tool, which would mean a lot of wasted CPU cycles copying data around

**Reply by Avi-D-coder** [reply]
@ipetkov it is a code smell, the reference says not to do it, but it's very common. Could we get an option to build specific dependencies by copying it to tmp, so that we can easily work around this issue without needing upstream changes.

https://doc.rust-lang.org/cargo/reference/build-script-examples.html#code-generation
> In general, build scripts should not modify any files outside of OUT_DIR. It may seem fine on the first blush, but it does cause problems when you use such crate as a dependency, because there's an implicit invariant that sources in .cargo/registry should be immutable. cargo won't allow such scripts when packaging.

**Reply by ipetkov** [reply]
> Could we get an option to build specific dependencies by copying it to tmp, so that we can easily work around this issue without needing upstream changes.

Unfortunately I'd be inclined to say no since that would be a can of worms to implement and maintain properly. We can't just check the target of what is being compiled, but in this case detect if such a dependency is present _anywhere_ in the closure. We also can't just build one specific crate in a mutable directory, the entire vendor tree would have to be moved to a mutable directory which complicates things further

**Reply by Avi-D-coder** [reply]
Ah shoot I did not think about that. What about supporting the dream2nix move everything method as a option?

**Reply by ipetkov** [reply]
I'm not sure it's necessary to support it as a first-class configuration option, though I'm not opposed to documenting the workaround: https://github.com/ipetkov/crane/pull/211

**Reply by Avi-D-coder** [reply]
@ipetkov adding the post patch from that FAQ PR breaks vendoring. Cargo can not longer find any of my dependencies.

```
postPatch = ''
  mkdir -p "$TMPDIR/nix-vendor"
  cp -r "$cargoVendorDir" "$TMPDIR/nix-vendor"
  chmod -R +w "$TMPDIR/nix-vendor"
  cargoVendorDir="$TMPDIR/nix-vendor"
'';
```


```
error: builder for '/nix/store/y1mpal3qy7sxhgmd7mrwbv9kl18kmyg9-local-model-data-server-deps-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       > configuring
       > setting source replacement config in /build/px7hiicilrq4m08xpkldxvrw9fhjr6f3-source/.cargo-home/config.toml using vendored directory /build/nix-vendor
       > default configurePhase, nothing to do
       > building
       > ++ command cargo --version
       > cargo 1.67.0-nightly (f6e737b1e 2022-12-02)
       > ++ command cargo check --profile release --all-targets
       > error: no matching package named `anyhow` found
       > location searched: registry `crates-io`
       > required by package `local-model-data-server v0.1.0 (/build/px7hiicilrq4m08xpkldxvrw9fhjr6f3-source)`
       For full logs, run 'nix log /nix/store/y1mpal3qy7sxhgmd7mrwbv9kl18kmyg9-local-model-data-server-deps-0.1.0.drv'.
error: 1 dependencies of derivation '/nix/store/r2yf1bmfh5n1czli4jv7gb4n0x9y2k1h-local-model-data-server-0.1.0.drv' failed to build
```

**Reply by ipetkov** [reply]
My bad, that should be `cp -r "$cargoVendorDir" -T "$TMPDIR/nix-vendor"` to avoid the nested directory when copying to `$TMPDIR/nix-vendor` I'll have a fix up in a bit

**Reply by ipetkov** [reply]
Fix is in https://github.com/ipetkov/crane/pull/214

**Reply by Avi-D-coder** [reply]
@ipetkov I tried it, and it's still not working. I'm running into a similar issue with `torch-sys` wanting to download in `build.rs`.

```
error: builder for '/nix/store/c18a5ccn1mippqvzqqmkib68lcvlzw20-local-model-data-server-deps-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       >   cargo:rerun-if-changed=/nix/store/nasgvbblxi25r6y1s82il4m0m4ghib68-glibc-2.35-224-dev/include/bits/types.h
       >   cargo:rerun-if-changed=/nix/store/1kcyy4z7lqz3f1mgmld0yc07fglvv4w0-clang-wrapper-11.1.0/resource-root/include/stdbool.h
       >   gcc  -I.              -O3 -std=c11   -fPIC -pthread -mavx -mavx2 -mfma -mf16c   -c ggml.c -o ggml.o
       >
       >   --- stderr
       >   Assembler messages:
       >   Fatal error: can't create ggml.o: Permission denied
       >   make: *** [Makefile:145: ggml.o] Error 1
       >   thread 'main' panicked at 'Failed to build libwhisper.a', /nix/store/m8xvl3b5qq3agq1dk8bq8rjqr7s15glv-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/whisper-rs-sys-0.2.1/build.rs:55:9
       >   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
       For full logs, run 'nix log /nix/store/c18a5ccn1mippqvzqqmkib68lcvlzw20-local-model-data-server-deps-0.1.0.drv'.
error: 1 dependencies of derivation '/nix/store/27i1pxs6l2abn59d8773jipv2w80d4ij-local-model-data-server-0.1.0.drv' failed to build
```

```
error: builder for '/nix/store/zngmfj91wsm67az73cbv6v9f8machxya-llms-deps-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       > Caused by:
       >   process didn't exit successfully: `/build/68rm40364nwkabr65bbl7rh6yvbfxbm3-source/target/release/build/torch-sys-a18fb4cad92325bd/build-script-build` (exit status: 101)
       >   --- stdout
       >   cargo:rerun-if-env-changed=TORCH_CUDA_VERSION
       >   cargo:rerun-if-env-changed=LIBTORCH
       >
       >   --- stderr
       >   thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: [6] Couldn't resolve host name (Could not resolve host: download.pytorch.org)', /nix/store/ym97izcwzpmadgvk7hxdrhgfx35mpahw-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/torch-sys-0.10.0/build.rs:142:48
       >   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/zngmfj91wsm67az73cbv6v9f8machxya-llms-deps-0.1.0.drv'.
error: 1 dependencies of derivation '/nix/store/m7m145qvh6h8f9h8yqq7w36wks6n78v3-llms-0.1.0.drv' failed to build

```

**Reply by ipetkov** [reply]
You may need to point `torch-sys` to a precompiled version of `libtorch` instead of having it try to compile its own. I'd recommend consulting their documentation for the appropriate way to do that (e.g. set the right environment variables)

**Reply by Avi-D-coder** [reply]
That's sorta the crux of the issue, dealing with setting many sys style crates to use libs compiled by nix is a pain.
Not every library support's it without upstream changes, at some point it becomes supporting nix/crane with Rust rather than crane supporting Rust. There should be some sort of easy workaround in crane for supporting sys style crates.

**Reply by ipetkov** [reply]
> There should be some sort of easy workaround in crane for supporting sys style crates.

What does an "easy workaround" actually look like?

>  dealing with setting many sys style crates to use libs compiled by nix is a pain.
Not every library support's it without upstream changes, at some point it becomes supporting nix/crane with Rust rather than crane supporting Rust

I understand it can be frustrating getting external projects to play nicely with Nix but I disagree with the characterization here. Crane fully supports building cargo projects. Cargo itself knows _nothing_ about system dependencies, besides doing whatever it is instructed by build scripts. I'm open to exploring ideas but at the end of the day there is only so little that we can do or guess downsteam of Nix and its sandbox, and downstream of whatever `build.rs` logic is trying to run.

There's _plenty_ of examples of `*-sys` crates which can build just fine under Nix. I know for a fact the following work as I use them in multiple projects which I build and deploy exclusively with Nix:
- `openssl-sys` just needs `openssl` and `pkg-config` listed as inputs and it can figure out the rest
- `libsqlite3-sys` is fully able to compile its own vendored source code with no additional inputs besides a C compiler being present

If a build script insists on touching the network at build time, it cannot be Crane's responsibility to guess and know how to satisfy it while running within the sandbox itself.

If a build script insists on expecting a FHS hierarchy, it cannot be Crane's responsibility to simulate a filesystem and guess what dependencies the build script expects to find where.

If a build script insists on running tools like `apt-get`, `dnf`, `yum`, or `brew`, it cannot be Crane's responsibility to emulate those tools and guess what they ought to be doing.

At the end of the day, it cannot be Crane's responsibility to automagically satisfy the requirements of any arbitrary build script under the sun without any additional intervention from a human who understands the project and its dependencies and how to satisfy them accordingly, especially when said build scripts expect full reign of the system which Nix explicitly takes away.
